k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java

```diff
@@ -447,4 +467,116 @@ public MessageBuilder setDraft(boolean isDraft) {
+     * @param sign turn on signature, if false other parameters will be ignored
+     * @param signIntent Intent to use to sign the message. This Intent must be
+     *                      fully functional and mustn't requires further interaction. If it isn't the case,
+     *                      message building will fail.
+     * @param api Link to an initialized OpenPgpApi
+     * @return this
+     */
+    public MessageBuilder setPgpMimeSignature(boolean sign, Intent signIntent, OpenPgpApi api){
+        if (sign) {
+            this.pgpMimeSignIntent = signIntent;
+            this.openPgpApi = api;
+        } else {
+            this.pgpMimeSignIntent = null;
+        }
+        return this;
+    }
+
+    /**
+     * Process the message in order to encapsulate it in a multipart/signed message.
+     * @see <a href="http://tools.ietf.org/html/rfc2015">RFC-2015</a>
+     * @param mime messsage to process
+     * @throws MessagingException
+     */
+    private void encapsulateMimeInMultipartSigned(MimeMessage mime) throws MessagingException, OpenPgpApiException {
+        /*
+         * Once set to true, text messages will be sign safe (RFC-3156 §3) until K9Mail is stopped.
+         * This "global parameter" is made to avoid a double generation (regular/sign safe) on the
+         * whole Part/Body hierarchy and still limit the scope of this extra encoding to pgp/mime users
+         * Downside it that even unsigned messages will contain extra-encoding once a message has
+         * been signed. But it will be transparent if the recipient decodes properly quoted-printable.
+         */
+        TextBody.setSignSafe(true);
+        MimeBodyPart bodyPart = mime.toBodyPart();
+        bodyPart.setUsing7bitTransport();
+        ByteArrayOutputStream messageToSign = new ByteArrayOutputStream();
+        try {
+            bodyPart.writeTo(messageToSign);
+            Intent result = openPgpApi.executeApi(pgpMimeSignIntent, new ByteArrayInputStream(messageToSign.toByteArray()), null);
+            if (result.getIntExtra(OpenPgpApi.RESULT_CODE, OpenPgpApi.RESULT_CODE_ERROR) != OpenPgpApi.RESULT_CODE_SUCCESS) {
+                OpenPgpError error = result.getParcelableExtra(OpenPgpApi.RESULT_ERROR);
+                Log.e(OpenPgpApi.TAG, error.getMessage());
+                throw new OpenPgpApiException(error);
+            }
+            byte[] signedData = result.getByteArrayExtra(OpenPgpApi.RESULT_DETACHED_SIGNATURE);
+
+            MimeMultipart multipartSigned = new MimeMultipart();
+            multipartSigned.setSubType("signed");
+            multipartSigned.addBodyPart(bodyPart);
+            multipartSigned.addBodyPart(new MimeBodyPart(new BinaryMemoryBody(signedData, MimeUtil.ENC_7BIT), "application/pgp-signature"));
+
+            MimeMessageHelper.setBody(mime, multipartSigned);
+            mime.addContentTypeParameter("protocol", "\"application/pgp-signature\"");
+            mime.addContentTypeParameter("micalg", "pgp-sha256");
```

Comment:
Does OpenKeyChain guarantee that SHA-256 will always be used to sign? The [DEFAULT_HASH_ALGORITHM value in PgpSecurityConstants](https://github.com/open-keychain/open-keychain/blob/master/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/PgpSecurityConstants.java) led me to believe that SHA-512 was the guaranteed default.